fix(flags): Add periodic cleanup for rate limiter memory leak

haacked opened the pull request:
## Problem

The `governor` crate uses a `DashMap` (like a `HashMap` but thread safe) to store tokens used for rate limiting. But it doesn't automatically clean them up, leading to potential memory leaks.

<img width="480" height="463" alt="image" src="https://github.com/user-attachments/assets/df63fd07-3e1c-4f16-83a0-9a12810975ae" />
<img width="1905" height="1510" alt="image" src="https://github.com/user-attachments/assets/122c7cbb-1eeb-46ac-bf0a-0e169cdb9f4b" />


## Changes

- Add `retain_recent()`, `shrink_to_fit()`, `len()`, and `is_empty()` methods to `FlagsRateLimiter`, `IpRateLimiter`, and `FlagDefinitionsRateLimiter`
- Add `spawn_rate_limiter_cleanup_task()` background task that periodically:
  - Calls `retain_recent()` to remove stale entries (keys that haven't been used within the rate limit window)
  - Calls `shrink_to_fit()` to reclaim memory
  - Emits Prometheus gauges (`flags_rate_limiter_token_entries`, `flags_rate_limiter_ip_entries`, `flags_rate_limiter_definitions_entries`) for monitoring
- Add `RATE_LIMITER_CLEANUP_INTERVAL_SECS` config option (default: 60 seconds)
- Add `hammer_rate_limiter.sh` test script

## How did you test this code?

Added a `hammer_rate_limiter.sh` script to hammer the endpoint for testing:

Make sure you're in the `rust/feature-flags` directory:

Terminal 1 - Start the server with a short cleanup interval for testing:

```bash
FLAGS_RATE_LIMIT_ENABLED=true FLAGS_IP_RATE_LIMIT_ENABLED=true \
RATE_LIMITER_CLEANUP_INTERVAL_SECS=5 \
RUST_LOG=debug cargo run 2>&1 | grep -E "(cleanup|entries|Rate limiter)"
```

Terminal 2 - Run this script:

```bash
./scripts/hammer_rate_limiter.sh
```

Expected output in Terminal 1:
- token_entries increases during the hammer run (e.g., 20-100+ entries)
- After requests stop, token_entries drops to 0 on the next cleanup cycle
- ip_entries=1 (all requests from localhost share one IP)

Example output:

```
2025-12-23T18:33:25.390214Z DEBUG feature_flags::router: Rate limiter cleanup completed token_entries=0 ip_entries=0 definitions_entries=0
2025-12-23T18:33:32.389444Z DEBUG feature_flags::router: Rate limiter cleanup completed token_entries=31 ip_entries=1 definitions_entries=0
2025-12-23T18:33:39.389600Z DEBUG feature_flags::router: Rate limiter cleanup completed token_entries=28 ip_entries=1 definitions_entries=0
2025-12-23T18:33:46.389849Z DEBUG feature_flags::router: Rate limiter cleanup completed token_entries=24 ip_entries=1 definitions_entries=0
2025-12-23T18:33:53.391285Z DEBUG feature_flags::router: Rate limiter cleanup completed token_entries=0 ip_entries=1 definitions_entries=0
```

Another test run where we let more tokens accumulate:

```bash
FLAGS_RATE_LIMIT_ENABLED=true FLAGS_IP_RATE_LIMIT_ENABLED=true \
FLAGS_BUCKET_REPLENISH_RATE=0.1 \
RATE_LIMITER_CLEANUP_INTERVAL_SECS=5 \
RUST_LOG=debug cargo run 2>&1 | grep -E "(cleanup|entries|Rate limiter)"
2025-12-23T18:41:43.287050Z DEBUG ThreadId(06) feature_flags::cohorts::cohort_cache_manager: Cohort cache metrics - size: 0 bytes, entries: 0
2025-12-23T18:41:43.289550Z DEBUG ThreadId(13) feature_flags::router: Rate limiter cleanup completed token_entries=0 ip_entries=0 definitions_entries=0
2025-12-23T18:41:49.289221Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=143 ip_entries=1 definitions_entries=0
2025-12-23T18:41:55.289510Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=424 ip_entries=1 definitions_entries=0
2025-12-23T18:42:01.290758Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=684 ip_entries=1 definitions_entries=0
2025-12-23T18:42:07.290153Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=936 ip_entries=1 definitions_entries=0
2025-12-23T18:42:13.287858Z DEBUG ThreadId(06) feature_flags::cohorts::cohort_cache_manager: Cohort cache metrics - size: 0 bytes, entries: 0
2025-12-23T18:42:13.290411Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=1 definitions_entries=0
2025-12-23T18:42:19.290096Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=1 definitions_entries=0
2025-12-23T18:42:25.291065Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=1 definitions_entries=0
2025-12-23T18:42:31.290810Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=0 definitions_entries=0
2025-12-23T18:42:37.291136Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=0 definitions_entries=0
2025-12-23T18:42:43.287673Z DEBUG ThreadId(05) feature_flags::cohorts::cohort_cache_manager: Cohort cache metrics - size: 0 bytes, entries: 0
2025-12-23T18:42:43.290186Z DEBUG ThreadId(06) feature_flags::router: Rate limiter cleanup completed token_entries=1000 ip_entries=0 definitions_entries=0
2025-12-23T18:42:49.291748Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=904 ip_entries=0 definitions_entries=0
2025-12-23T18:42:55.291010Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=620 ip_entries=0 definitions_entries=0
2025-12-23T18:43:01.290960Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=357 ip_entries=0 definitions_entries=0
2025-12-23T18:43:07.290838Z DEBUG ThreadId(08) feature_flags::router: Rate limiter cleanup completed token_entries=105 ip_entries=0 definitions_entries=0
2025-12-23T18:43:13.288481Z DEBUG ThreadId(04) feature_flags::cohorts::cohort_cache_manager: Cohort cache metrics - size: 0 bytes, entries: 0
2025-12-23T18:43:13.289916Z DEBUG ThreadId(04) feature_flags::router: Rate limiter cleanup completed token_entries=0 ip_entries=0 definitions_entries=0
```


👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changelog

No - infrastructure fix, not user-facing.

haacked committed:
Add periodic cleanup for rate limiter memory leak

The governor crate's keyed rate limiters store per-key state in a `DashMap` that grows unboundedly without periodic cleanup. This was causing memory to grow linearly over time (observed ~2GB growth over 7 days).

Add a background task that periodically calls `retain_recent()` and
`shrink_to_fit()` on all rate limiters to remove stale entries. This follows the pattern recommended by `tower_governor`.

Changes:
- Add `retain_recent()`, `shrink_to_fit()`, `len()` methods to all rate limiters
- Spawn cleanup task in router.rs (runs every 60 seconds by default)
- Add `RATE_LIMITER_CLEANUP_INTERVAL_SECS` config option
- Add metrics: `flags_rate_limiter_token_entries`, `flags_rate_limiter_ip_entries`,
  `flags_rate_limiter_definitions_entries`

haacked committed:
Add jitter to rate limiter cleanup interval

Adds 0-10 seconds of random jitter to the cleanup interval to prevent
synchronized cleanup across replicas, which could cause brief latency
spikes fleet-wide.

This follows the same pattern used by OverflowLimiter in
rust/common/limiters/src/overflow.rs.

haacked committed:
Add script to test rate limiter cleanup

Adds a hammer script that generates many unique tokens and sends
requests to populate the rate limiter state. Useful for manually
verifying that the periodic cleanup correctly removes stale entries.

Usage instructions are included in the script header.

haacked committed:
Document is_empty() methods as approximate

The underlying DashMap may have concurrent modifications, so these
methods should be used for monitoring purposes only, not correctness.

haacked committed:
Add panic recovery to rate limiter cleanup task

If the cleanup logic panics, log the error and continue to the next
interval rather than letting the task die forever (which would cause
the memory leak to resume silently).

haacked committed:
Improve is_empty() implementation and add tests

- Add is_empty() to inner KeyedRateLimiter for proper delegation
- FlagDefinitionsRateLimiter.is_empty() now short-circuits instead of
  computing full len()
- Add tests for is_empty() behavior

gustavohstrassburger commented on rust/feature-flags/src/router.rs:
```diff
@@ -177,6 +192,59 @@ pub fn router(
+) {
+    tokio::spawn(async move {
+        // Add jitter to avoid synchronized cleanup across replicas
+        let jitter_secs = rand::thread_rng().gen_range(0..10);
```
If the jitter is really necessary and clashes should be avoided, we should consider if 10s is enough given that we are currently running 50 pods.

gustavohstrassburger commented on rust/feature-flags/src/api/flag_definitions_rate_limiter.rs:
```diff
@@ -163,6 +163,62 @@ where
+    /// Returns true if no keys are currently tracked in any limiter.
+    ///
+    /// Note: This is approximate and shares the same limitations as [`Self::len`].
+    pub fn is_empty(&self) -> bool {
```
I'm not that of a fan of adding methods that are only used in tests, specially since you could test pretty much the same simply by using the `len` method. 

This might be faster in some cases because of the `if !self.default_limiter.is_empty()`, but I'm not sure if this was the reason why.

gustavohstrassburger reviewed (approved):
Left some comments, but none are a blocker, just some thoughts.

haacked commented on rust/feature-flags/src/api/flag_definitions_rate_limiter.rs:
```diff
@@ -163,6 +163,62 @@ where
+    /// Shrinks the capacity of all rate limiter state stores if possible.
+    ///
+    /// Should be called after `retain_recent()` to reclaim memory.
+    pub fn shrink_to_fit(&self) {
```
Good point!

haacked committed:
Remove is_empty() methods that were only used in tests

Per code review feedback, removed is_empty() from rate limiter types
since they were only used in tests. Tests now use len() == 0 instead.

Added #[allow(clippy::len_without_is_empty)] since these types
intentionally only expose len() for metrics purposes.

haacked committed:
Remove unnecessary jitter from rate limiter cleanup

The cleanup task operates on local in-memory data only - no shared
resources between pods. Jitter is unnecessary since synchronized
cleanup across pods doesn't cause any contention.

haacked committed:
Add cleanup() method to rate limiters

Per code review feedback, added a cleanup() convenience method that
combines retain_recent() and shrink_to_fit() since they're always
called together. This simplifies the call site in router.rs.

haacked committed:
Simplify rate limiter API by consolidating cleanup methods

Remove public retain_recent() and shrink_to_fit() methods from all
rate limiters since they're always called together. The cleanup()
method now calls the inner implementations directly, reducing API
surface and code duplication.

Addresses review feedback from @gvanderlinden.

haacked commented on rust/feature-flags/src/api/flags_rate_limiter.rs:
```diff
@@ -122,6 +122,38 @@ impl KeyedRateLimiter {
+    /// entries for keys that are no longer actively making requests.
+    fn retain_recent(&self) {
+        self.limiter.retain_recent();
+    }
```
We're going to trust the library to have tests of its own functionality.

haacked committed:
Fix macOS portability in hammer_rate_limiter.sh

Replace `date +%s%N | md5sum` with `openssl rand -hex 4` for generating
unique tokens. The previous approach used GNU-specific features:
- %N (nanoseconds) isn't supported by BSD date on macOS
- md5sum command doesn't exist on macOS (it's md5)

openssl is available on both Linux and macOS.

haacked committed:
Improve hammer_rate_limiter.sh output and IP simulation

- Suppress interleaved status codes during execution
- Add response code summary at the end
- Generate random X-Forwarded-For IPs to test IP-based rate limiting

haacked committed:
Remove unused response code tracking from hammer script
